Add manual datastore refresh endpoint

rainest opened the pull request:
### Summary

Add manual datastore refresh functionality.

### Full changelog

* Adds refresh functions for Cassandra and Postgres. The latter is a dummy function for compatibility.
* Adds Admin API route to call datastore refresh functions.

### Background

Related to issue #2674. Although this is not a complete fix (ideally refreshes should not require manual intervention) it's a better workaround than completely restarting a node.

For Cassandra, this adds a simple means of calling the existing Cluster:refresh() function provided by the lua-cassandra driver. Postgres does not have similar needs, so its function is a dummy that the Admin API route can always call successfully.

No tests added as lua-cassandra should handle everything important and test it independently. I performed basic testing by starting Kong, adding a node to its Cassandra cluster, calling refresh, and confirming that the new node's IP was selected by the DB load balancing policy. I then removed the node from the cluster, observed that `set_peer_down()` was called on the node IP, refreshed again, and confirmed that calls to `set_peer_down()` had stopped.

Commit:
feat(dao) add DB refresh functions

Adds refresh functions to the Cassandra and Postgres interfaces.
The Cassandra function calls lua-cassandra's Cluster:refresh function,
which updates the list of nodes in the cluster.
The Postgres function does nothing, and only serves to provide a
consistent interface for both datastores.

Related to #2674

Commit:
feat(admin) add /refresh route

Adds the /refresh route to the Admin API. A PUT to this route calls the
datastore's refresh method. If successful, this returns a 204. If
unsuccessful, this returns a 503 with the datastore driver's error in
the body.

Related to #2674

hishamhm commented on kong/dao/db/postgres.lua:
```diff
@@ -631,4 +631,15 @@ function _M:reachable()
+    return nil, Errors.db(err)
+  end
+
+  return true
```
minor suggestion: no need to keep the "error handling template" in this Postgres implementation. In similar situations we just keep it simple, with the body keeping only `return true` and the comment.

Commit:
feat(dao) remove superfluous refresh code

rainest commented:
There seems to be an underlying issue in lua-cassandra that prevents this from working as expected. After refreshing, Kong's load balancing policy selects the correct nodes, but other actions will still attempt to use old addresses and fail. Reviewing how refresh modifies the driver's data structures to see why this occurs.

Commit:
feat(admin) broadcast database refresh to workers

Broadcast database refresh requests to all workers to ensure that
non-shared topology information is refreshed by all workers.

rainest commented:
Now improved with lua-resty-worker-events to run refresh on all workers. Previously, only one worker would run refresh, and other workers would continue to use outdated load balancer tables unsuccessfully.

While testing, I used [a modified version](https://gist.github.com/rainest/0a59608cf6c86d84b113c26b465b5419) of the lua-cassanda cluster object to litter debug logging statements everywhere and confirm that refresh was run on multiple workers. There are no functional changes there, however, so I have not made a separate PR against lua-cassandra.

thibaultcha commented on kong/runloop/handler.lua:
```diff
@@ -165,9 +165,16 @@ return {
+      worker_events.register(function()
+        local ok, err = dao.db:refresh()
+        if not ok then
+          ngx.log(ERR, "[events] failed to schedule DB refresh: ", err)
```
If the cluster topology fails to refresh, it warrants a higher logging level. That is because more than one transaction/request might be affected by this error. In fact, this error may trigger cascading failures in the long run. A logging level set to `ngx.CRIT` seems more appropriate here.

Also, the error message seems wrong: it isn't the scheduling that failed in this case, but the refresh itself. And the other way around: the message currently showed in the Admin API failure branch seems to be wrong; in this case, it is indeed the _scheduling_ that failed, and not the refresh itself.

thibaultcha reviewed (changes requested):
I have concerns about the use of `refresh()` at runtime, as detailed in the related comment. This method was not implemented as a complete solution back then, and was never completed to support true runtime topology refresh... Let me know your thoughts.

Commit:
feat(admin) correct refresh error content

Commit:
feat(dao) add DB refresh to new dao

Commit:
fix(dao) clear refresh lint issues

rainest commented on kong/dao/db/cassandra.lua:
```diff
@@ -834,4 +834,13 @@ function _M:reachable()
+  end
+
+  return true
+end
```
Updated to include equivalent functions in the new DAO and call both until the old DAO is removed.

angrosz commented:
When are you going to have the final solution of this problem? I have the same problem as described in #2674.

angrosz commented:
This solution is a manual action. Is it possible to implement an automatic cassandra cluster refresh?

Commit:
feat(dao) block C* queries while refreshing

Block attempts to query Cassandra while refreshing topology. Since host
metadata is cleared during a refresh, it will not be available to query jobs
until the refresh completes and repopulates them. Queries issued during
the blocking period will be queued and run after refresh completes.

rainest commented:
@thibaultcha added [a locking mechanism to block queries](https://github.com/Kong/kong/pull/3752/commits/3f23b2540bcf869a2d47e28643b46bb931e207d5) at the Kong level. I initially attempted a driver-level solution, but wasn't sure if there was a way to implement it at the [host query level](https://github.com/thibaultcha/lua-cassandra/blob/664422d2e58fbd7e7cc30e2f6a576fd518c171ce/lib/cassandra/init.lua#L485). Kong sends queries directly to hosts in some cases (see [this example](https://github.com/Kong/kong/blob/master/kong/db/strategies/cassandra/connector.lua#L193) in the new DAO versus a [query sent to the cluster](https://github.com/Kong/kong/blob/master/kong/dao/db/cassandra.lua#L316) in the old DAO) and I didn't see an existing link back to the cluster object (where refresh grabs its semaphore). I still have several open questions/issues:
* Should this use the existing `cassandra_timeout` or its own (currently hardcoded to 10s arbitrarily)?
* Should queries always attempt to grab the semaphore? I added a check to `count()` first to ensure they only did so if someone invokes `refresh()`, as otherwise queries could block each other, albeit only briefly (they only grab the resource for a preclearance check/do not hold onto it until the query completes). Checking via `count()` before running `wait()` means that there isn't a true guarantee of exclusivity, though it's a stronger safeguard than not checking anything.
* Is there a practical means of adding an automated test for this? I don't know of a means to guarantee a test query will attempt to run while `refresh()` holds the semaphore. I do see expected behavior with a [modified version](https://gist.github.com/rainest/777673a5edbfada6ff2c2795ed5a6a1b#file-connector-lua-L337-L338) that intentionally delays refresh completion enough that I can manually issue queries then.

@angrosz this should work as-is if you want to make a custom build that includes it, though it may cause some queries to fail during a brief period. The commit mentioned above should prevent those failures.

A fully-automated solution is possible but would require new functionality in the driver. We do not yet have support for handling [CQL EVENT messages](https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L729-L739). Alternately, adding a PostStart hook that calls the Kong refresh endpoint to Cassandra pods would likely handle automating this for reschedule or respawn events, though manual downscaling would still require a manual refresh.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=3752) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=3752) before we can accept your contribution.<br/><hr/>**Travis Raines** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=3752) it.</sub>

thibaultcha commented:
Closing in favor of https://github.com/Kong/kong/pull/5071
